src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,176 @@
+                GAS.Mint(engine, account, OracleRequestPrice);
+            }
+        }
+
+        [ContractMethod(OracleRequestPrice, CallFlags.AllowModifyStates)]
+        private void Request(ApplicationEngine engine, string url, string filter, string callback, StackItem userData, long gasForRepsonse)
+        {
+            //Check the arguments
+            if (Utility.StrictUTF8.GetByteCount(url) > MaxUrlLength
+                || (filter != null && Utility.StrictUTF8.GetByteCount(filter) > MaxFilterLength)
+                || Utility.StrictUTF8.GetByteCount(callback) > MaxCallbackLength
+                || gasForRepsonse <= 0)
+                throw new ArgumentException();
+
+            //Mint gas for the response
+            engine.AddGas(gasForRepsonse);
+            GAS.Mint(engine, Hash, gasForRepsonse);
+
+            //Increase the request id
+            StorageItem item_id = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_RequestId));
+            ulong id = BitConverter.ToUInt64(item_id.Value) + 1;
+            item_id.Value = BitConverter.GetBytes(id);
+
+            //Put the request to storage
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Request).Add(item_id.Value), new StorageItem(new OracleRequest
+            {
+                OriginalTxid = GetOriginalTxid(engine),
+                GasForResponse = gasForRepsonse,
+                Url = url,
+                Filter = filter,
+                CallbackContract = engine.CallingScriptHash,
+                CallbackMethod = callback,
+                UserData = BinarySerializer.Serialize(userData, MaxUserDataLength)
+            }));
+
+            //Add the id to the IdList
+            engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_IdList).Add(GetUrlHash(url)), () => new StorageItem(new IdList())).GetInteroperable<IdList>().Add(id);
+        }
+
+        [ContractMethod(0, CallFlags.AllowModifyStates)]
+        private void SetOracleNodes(ApplicationEngine engine, ECPoint[] nodes)
+        {
+            if (!CheckCommittee(engine)) throw new InvalidOperationException();
+            NodeList list = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_NodeList)).GetInteroperable<NodeList>();
+            list.Clear();
+            list.AddRange(nodes);
+        }
+
+        [ContractMethod(0_01000000, CallFlags.AllowStates)]
+        private bool Verify(ApplicationEngine engine)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            OracleResponse response = tx.Attributes.OfType<OracleResponse>().FirstOrDefault();
+            if (response is null) return false;
+            StorageKey key = CreateStorageKey(Prefix_Request).Add(response.Id);
+            OracleRequest request = engine.Snapshot.Storages.TryGet(key)?.GetInteroperable<OracleRequest>();
+            if (request is null) return false;
+            return tx.NetworkFee + tx.SystemFee <= request.GasForResponse;
```

Comment:
If the oracle request force to have one specific script, how can we withdraw the funds deposited during the Request in this contract?

Reply:
We can't. `GasForResponse` has the same logic with `SystemFee`, they can't be withdrawn.

Reply:
If we can move the code of `OracleResponse.Verify` here, then we don't need to add `Verify` method for TransactionAttribute?

Reply:
No. Attackers can bypass this verification by adding their own `sender`.

Reply:
Here, if `tx.NetworkFee + tx.SystemFee > request.GasForResponse`, the response tx is invalid.